resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/kubernetes/submit/v2/MountedDependencyManagerImpl.scala

```diff
@@ -0,0 +1,283 @@
+    }
+    // TODO provide credentials properly when the staging server monitors the Kubernetes API.
+    val kubernetesCredentialsString = OBJECT_MAPPER.writer()
+      .writeValueAsString(KubernetesCredentials(None, None, None, None))
+    val labelsAsString = OBJECT_MAPPER.writer().writeValueAsString(podLabels)
+
+    val filesRequestBody = RequestBody.create(
+      okhttp3.MediaType.parse(MediaType.MULTIPART_FORM_DATA), filesTgz)
+
+    val kubernetesCredentialsBody = RequestBody.create(
+      okhttp3.MediaType.parse(MediaType.APPLICATION_JSON), kubernetesCredentialsString)
+
+    val namespaceRequestBody = RequestBody.create(
+      okhttp3.MediaType.parse(MediaType.TEXT_PLAIN), podNamespace)
+
+    val labelsRequestBody = RequestBody.create(
+      okhttp3.MediaType.parse(MediaType.APPLICATION_JSON), labelsAsString)
+
+    val service = retrofitClientFactory.createRetrofitClient(
+      stagingServerUri,
+      classOf[ResourceStagingServiceRetrofit],
+      stagingServiceSslOptions)
+    val uploadResponse = service.uploadResources(
+      labelsRequestBody, namespaceRequestBody, filesRequestBody, kubernetesCredentialsBody)
+    val rawSecret = getTypedResponseResult(uploadResponse)
+    rawSecret
+  }
+
+  override def configurePodToMountLocalDependencies(
+      driverContainerName: String,
+      initContainerSecret: Secret,
+      initContainerConfigMap: ConfigMap,
+      originalPodSpec: PodBuilder): PodBuilder = {
+    val sharedVolumeMounts = Seq[VolumeMount](
+      new VolumeMountBuilder()
+        .withName(DOWNLOAD_JARS_VOLUME_NAME)
+        .withMountPath(jarsDownloadPath)
+        .build(),
+      new VolumeMountBuilder()
+        .withName(DOWNLOAD_FILES_VOLUME_NAME)
+        .withMountPath(filesDownloadPath)
+        .build())
+
+    val initContainers = Seq(new ContainerBuilder()
+      .withName("driver-init")
+      .withImage(initContainerImage)
+      .withImagePullPolicy("IfNotPresent")
+      .addNewVolumeMount()
+        .withName(INIT_CONTAINER_PROPERTIES_FILE_VOLUME)
+        .withMountPath(INIT_CONTAINER_PROPERTIES_FILE_MOUNT_PATH)
+        .endVolumeMount()
+      .addNewVolumeMount()
+        .withName(INIT_CONTAINER_SECRETS_VOLUME_NAME)
+        .withMountPath(INIT_CONTAINER_SECRETS_VOLUME_MOUNT_PATH)
+        .endVolumeMount()
+      .addToVolumeMounts(sharedVolumeMounts: _*)
+      .addToArgs(INIT_CONTAINER_PROPERTIES_FILE_PATH)
+      .build())
+
+    // Make sure we don't override any user-provided init containers by just appending ours to
+    // the existing list.
+    val resolvedInitContainers = originalPodSpec
+      .editMetadata()
+      .getAnnotations
+      .asScala
+      .get(INIT_CONTAINER_ANNOTATION)
+      .map { existingInitContainerAnnotation =>
+        val existingInitContainers = OBJECT_MAPPER.readValue(
+          existingInitContainerAnnotation, classOf[List[Container]])
+        existingInitContainers ++ initContainers
```

Comment:
here we add the Spark init container after the existing ones specified by the user.  Why after instead of before?  Does it need to be configurable?